File: linker/lib/address_parser.rb
require File.dirname(__FILE__) + '/parser'

class AddressParser < Parser
  attr_accessor :base_address
  
  def initialize(linker)
    @base_address = 0
    super(linker)
  end
   
  def parse
    while (@char = reader.next)
      detect_symbols_and_base_addresses
    end
    reader.file.rewind
  end

private

  def detect_symbols_and_base_addresses
    # symbols
    module_symbols = {}
    parse_number.times do |i|
      symbol           = parse_word
      if SymbolTable.defines?(symbol)
        SymbolTable.errors[symbol] = "Error: This variable is multiply defined; first value used."
      else
        module_symbols[symbol] = parse_number + @base_address
        SymbolTable.table[symbol] = module_symbols[symbol]
      end
    end
    
    # base_addresses
    skip_use_list
    skip_program(module_size = parse_number)
    
    MemoryMap.create_program_module(@base_address, module_symbols)
    
    @base_address += module_size
  end
  
  def skip_use_list
    parse_number.times { |i| parse_word }
  end
  
  def skip_program(count = 0)
    count.times do |i|
      parse_word
      parse_number
    end
  end
    
end

File: java/linker/src/lib/AddressParser.java
package lib;

import java.io.IOException;
import java.util.Hashtable;

/**
 * User: Ross Kaffenberger
 * Date: Sep 22, 2009
 * Time: 11:37:03 PM
 */
public class AddressParser extends Parser {
    private int baseAddress = 0;

    public AddressParser(Linker linker) {
        super(linker);
    }

    public void parse() throws IOException {
        readNextCharString();
        while (!currentCharString().equals("")) {
            detectSymbolsAndBaseAddresses();
        }

        getReader().rewind();
    }

    private void detectSymbolsAndBaseAddresses() throws IOException {
        Hashtable<String, Integer> moduleSymbols = new Hashtable<String, Integer>();
        int nd = parseNumber();
        for (int i = 1; i <= nd; i++) {
            String symbol = parseWord();
            if (SymbolTable.defines(symbol)) {
                String error = "Error: This variable is multiply defined; first value used.";
                SymbolTable.getInstance().getErrors().put(symbol, error);
            } else {
                moduleSymbols.put(symbol, parseNumber() + this.baseAddress);
                SymbolTable.getInstance().put(symbol, moduleSymbols.get(symbol));
            }
        }
        skipUseList();
        int moduleSize = parseNumber();
        skipProgram(moduleSize);
        MemoryMap.createProgramModule(this.baseAddress, moduleSymbols);
        this.baseAddress += moduleSize;
        skipWhiteSpace();
    }

    private void skipProgram(int moduleSize) throws IOException {
        for (int i = 1; i <= moduleSize; i++) {
            parseWord();
            parseNumber();
        }
    }

    private void skipUseList() throws IOException {
        int nu = parseNumber();
        for (int i = 1; i <= nu; i++) {
            parseWord();
        }

    }

}
